ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/block/ReexecutingExecutionPayloadBlockManager.java

```diff
@@ -0,0 +1,172 @@
+            });
+  }
+
+  @Override
+  public void onSlot(UInt64 slot) {
+    super.onSlot(slot);
+    pendingBlocks.removeIf(block -> block.getSlot().isLessThan(slot.plus(EXPIRES_IN_SLOT)));
+  }
+
+  @Override
+  public SafeFuture<BlockImportResult> importBlock(SignedBeaconBlock block) {
+    return super.importBlock(block)
+        .thenPeek(
+            blockImportResult -> {
+              if (requiresReexecution(blockImportResult)) {
+                pendingBlocks.add(block);
+              }
+            });
+  }
+
+  private void execute() {
+    if (reexecutingExecutionPayload.compareAndSet(false, true)) {
+      LOG.debug("re-executing execution payloads for queued blocks");
+      SafeFuture.allOfFailFast(
+              pendingBlocks.stream()
+                  .map(this::reexecuteExecutionPayload)
+                  .toArray(SafeFuture[]::new))
+          .thenPeek(__ -> reexecutingExecutionPayload.set(false))
+          .exceptionally(
+              error -> {
+                if (error instanceof StopExecutionError) {
+                  LOG.error(
+                      "An error occurred which prevents additional payload executions: {} reason: {}",
+                      error.getMessage(),
+                      error.getCause().getMessage());
+                  return null;
+                }
+                throw new RuntimeException(error);
+              })
+          .reportExceptions();
+    }
+  }
+
+  private SafeFuture<Void> reexecuteExecutionPayload(final SignedBeaconBlock block) {
+    return super.importBlock(block)
+        .thenAccept(
+            blockImportResult -> {
+              if (canBeRemoved(blockImportResult)) {
+                pendingBlocks.remove(block);
+              }
+              applySubsequentExecutionPreventionCheck(blockImportResult);
+            });
+  }
+
+  private boolean requiresReexecution(BlockImportResult blockImportResult) {
+    final FailureReason failureReason = blockImportResult.getFailureReason();
+    return failureReason == FailureReason.FAILED_EXECUTION_PAYLOAD_EXECUTION
+        || failureReason == FailureReason.FAILED_EXECUTION_PAYLOAD_EXECUTION_SYNCING;
+  }
+
+  private boolean canBeRemoved(BlockImportResult blockImportResult) {
+    if (blockImportResult.isSuccessful()) {
+      return true;
+    }
+    final FailureReason failureReason = blockImportResult.getFailureReason();
+    return failureReason != FailureReason.FAILED_EXECUTION_PAYLOAD_EXECUTION
+        && failureReason != FailureReason.FAILED_EXECUTION_PAYLOAD_EXECUTION_SYNCING;
+  }
+
+  private void applySubsequentExecutionPreventionCheck(BlockImportResult blockImportResult) {
+    final FailureReason failureReason = blockImportResult.getFailureReason();
+    if (failureReason == FailureReason.FAILED_EXECUTION_PAYLOAD_EXECUTION
+        || failureReason == FailureReason.FAILED_EXECUTION_PAYLOAD_EXECUTION_SYNCING) {
+      throw new StopExecutionError(blockImportResult);
+    }
```

Comment:
I'm not sure this is right.  We might get a `SYNCING` response from the EL for one fork but be able to import blocks from another fork. One way that can happen is if an attacker publishes the merge block (first to include a non-default payload) that references a non-existent parent.  We'll always get a syncing response for that block but would be able to import other blocks.

I think it's also possible that the EL might encounter an error processing one block but be fine with others (though that's likely an EL bug but could be a block designed to use excessive memory or something like that). So even the non-syncing failure probably shouldn't stop us trying to execute other blocks.